File: projects/day-001-mern-rate-limiter/src/RateLimiter.js
/**
 * Token Bucket Rate Limiter
 * High-performance, in-memory rate limiter with fractional token refill
 * and RFC 6585 compliance.
 */

class TokenBucket {
  /**
   * @param {number} capacity - Maximum burst capacity of tokens
   * @param {number} refillRate - Tokens added per second
   */
  constructor(capacity, refillRate) {
    if (capacity <= 0 || refillRate <= 0) {
      throw new Error('Capacity and refill rate must be positive numbers');
    }
    this.capacity = capacity;
    this.refillRate = refillRate;
    this.tokens = capacity;
    this.lastRefill = Date.now();
  }

  /**
   * Replenishes tokens based on the elapsed wall-clock time.
   * @private
   */
  _replenish() {
    const now = Date.now();
    const elapsedSeconds = (now - this.lastRefill) / 1000;
    const addedTokens = elapsedSeconds * this.refillRate;
    this.tokens = Math.min(this.capacity, this.tokens + addedTokens);
    this.lastRefill = now;
  }

  /**
   * Attempts to consume tokens.
   * @param {number} tokens - Number of tokens to acquire
   * @returns {{ allowed: boolean, headers: Object }}
   */
  acquire(tokens = 1) {
    this._replenish();

    if (this.tokens >= tokens) {
      this.tokens -= tokens;
      return {
        allowed: true,
        headers: {
          'X-RateLimit-Limit': this.capacity,
          'X-RateLimit-Remaining': Math.floor(this.tokens),
          'X-RateLimit-Reset': Math.ceil((this.lastRefill + 1000) / 1000)
        }
      };
    }

    const needed = tokens - this.tokens;
    const retryAfter = Math.ceil(needed / this.refillRate);

    return {
      allowed: false,
      headers: {
        'X-RateLimit-Limit': this.capacity,
        'X-RateLimit-Remaining': 0,
        'Retry-After': retryAfter
      }
    };
  }

  /**
   * Connects as standard Express / Node.js middleware.
   * @returns {Function}
   */
  middleware() {
    return (req, res, next) => {
      const result = this.acquire(1);
      Object.entries(result.headers).forEach(([key, val]) => {
        res.setHeader(key, val);
      });

      if (result.allowed) {
        return next();
      }

      return res.status(429).json({
        error: 'Too Many Requests',
        message: 'Rate limit exceeded. Please retry later.',
        retryAfter: result.headers['Retry-After']
      });
    };
  }
}

module.exports = TokenBucket;

File: projects/day-008-python-rate-limiter/src/rate_limiter.py
"""
Token Bucket Rate Limiter
A high-precision, thread-safe rate limiter implementing fractional token refill.
"""

import time
import threading
from typing import Tuple, Dict, Any, Optional

class TokenBucket:
    """
    Thread-safe Token Bucket Rate Limiter.
    Allows bursts up to `capacity` tokens, replenishing at `refill_rate` tokens per second.
    """
    def __init__(self, capacity: int, refill_rate: float):
        if capacity <= 0 or refill_rate <= 0:
            raise ValueError("Capacity and refill rate must be positive values")
        self.capacity = float(capacity)
        self.refill_rate = float(refill_rate)
        self.tokens = float(capacity)
        self.last_refill = time.monotonic()
        self._lock = threading.Lock()

    def _refill(self, now: float) -> None:
        elapsed = now - self.last_refill
        refill_amount = elapsed * self.refill_rate
        self.tokens = min(self.capacity, self.tokens + refill_amount)
        self.last_refill = now

    def acquire(self, tokens: int = 1) -> Tuple[bool, Dict[str, Any]]:
        """
        Attempts to acquire `tokens`.
        Returns a tuple of (allowed: bool, headers: dict) conforming to RFC 6585.
        """
        with self._lock:
            now = time.monotonic()
            self._refill(now)

            if self.tokens >= tokens:
                self.tokens -= tokens
                return True, {
                    "X-RateLimit-Limit": int(self.capacity),
                    "X-RateLimit-Remaining": int(self.tokens),
                    "X-RateLimit-Reset": int(self.last_refill + (1.0 / self.refill_rate))
                }
            
            # Calculate wait time until enough tokens are replenished
            needed = tokens - self.tokens
            retry_after = needed / self.refill_rate
            return False, {
                "X-RateLimit-Limit": int(self.capacity),
                "X-RateLimit-Remaining": 0,
                "Retry-After": round(retry_after, 2)
            }
